torch/csrc/jit/passes/bailout_graph.cpp

```diff
@@ -0,0 +1,216 @@
+    auto node = old_value->node();
+    // this reduces the number of inputs to a bailout graph significantly
+    // making it easier to debug
+    if (node->kind() == prim::Constant) {
+      auto new_const = copy_graph_->createClone(node, {nullptr});
+      copy_graph_->block()->appendNode(new_const);
+      return new_const->output();
+    }
+
+    live_inputs_.push_back(old_value);
+    auto new_value = copy_graph_->block()->addInput();
+    return mapExistingInputForValue(old_value, new_value);
+  }
+
+  Value* mapExistingInputForValue(Value* old_value, Value* new_value) {
+    this->old_to_new_[old_value] = new_value;
+    new_value->copyMetadata(old_value);
+    return new_value;
+  }
+
+  Value* getOrAddInputForValue(Value* v) {
+    if (this->old_to_new_.count(v) == 0) {
+      return addNewInputForValue(v);
+    } else {
+      return this->old_to_new_[v];
+    }
+  }
+
+  // buildBailOutBlockFrom builds a bailout graph from
+  // a given node `n` until the end of the owning block
+  // If `n` belongs to `prim::If` or `prim::Loop`
+  // buildBailOutGraphForOwningNode continues
+  // from block's owning node (e.g. `prim::If` or
+  // `prim::Loop`)
+  void buildBailOutBlockFrom(Node* n) {
+    auto* block = copy_graph_->block();
+    auto b = n->owningBlock();
+    graph_node_list_iterator it(n, kNextDirection);
+    for (; it != b->nodes().end(); it++) {
+      auto env = [this](Value* v) { return getOrAddInputForValue(v); };
+
+      auto node = *it;
+
+      auto new_node = block->appendNode(copy_graph_->createClone(node, env));
+      for (size_t i = 0; i < node->outputs().size(); ++i) {
+        auto oo = node->outputs()[i];
+        auto no = new_node->outputs()[i];
+        old_to_new_[oo] = no;
+        no->copyMetadata(oo);
+      }
+    }
+
+    // we are either in `prim::If` or `prim::Loop`
+    // bailout graph building will continue from `outer_node` next
+    // remember/map the outputs needed to unwind this `prim::If`
+    // or `prim::Loop`
+    auto outer_node = n->owningBlock()->owningNode();
+    if (outer_node) {
+      buildBailOutGraphForOwningNode(n, outer_node);
+    }
+  }
+
+  // buildBailOutGraphForOwningNode builds a bailout graph
+  // for a CFG given node () as follows:
+  // * if `inner_node` is in a true or false arm of an `prim::If` node
+  // the rest of the arm gets inlined in the bailout graph
+  // and we continue building the rest of the bailout graph
+  // from the node *following* `outer_node`.
+  // * if `inner_node` is in a loop, the rest of the loop gets inlined
+  // we remember the loop outputs needed to execute the
+  // we adjust the upper limit (old_upper_limit - current_iteration)
+  // note the stop condition is already mapped (remembered)
+  // then, we continue building the rest of the bailout graph
+  // starting from the owning `outer_node`
+  // which implies `outer_node` will be cloned
+  void buildBailOutGraphForOwningNode(Node* inner_node, Node* outer_node) {
```

Comment:
Function signature is weird. inner_node is only used to get block outputs. Suggest:

`buildBailOutGraphForOwningNode(Node* outer_node, at::ArrayRef<Value*> block_outputs)`

Reply:
Also it would make more sense for those block outputs to be passed through the old_to_new_ map _before_ they reach this function so that the represent the values in the new graph. The old values are not used by this function.